storage/adapter.go

```diff
@@ -0,0 +1,407 @@
+package storage
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"math"
+
+	"github.com/filecoin-project/go-address"
+	"github.com/filecoin-project/go-sectorbuilder"
+	s2 "github.com/filecoin-project/go-storage-miner"
+	"github.com/ipfs/go-cid"
+	"golang.org/x/xerrors"
+
+	"github.com/filecoin-project/lotus/build"
+	"github.com/filecoin-project/lotus/chain/actors"
+	"github.com/filecoin-project/lotus/chain/events"
+	"github.com/filecoin-project/lotus/chain/types"
+)
+
+type TicketFn func(context.Context) (*sectorbuilder.SealTicket, error)
+
+type StorageMinerNodeAdapter struct {
+	api    storageMinerApi
+	events *events.Events
+
+	maddr address.Address
+	waddr address.Address
+
+	tktFn TicketFn
+}
+
+func NewStorageMinerNodeAdapter(n storageMinerApi, evts *events.Events, maddr, waddr address.Address, tktFn TicketFn) *StorageMinerNodeAdapter {
+	return &StorageMinerNodeAdapter{api: n, events: evts, tktFn: tktFn, maddr: maddr, waddr: waddr}
+}
+
+func (m *StorageMinerNodeAdapter) SendSelfDeals(ctx context.Context, pieces ...s2.PieceInfo) (cid.Cid, error) {
+	deals := make([]actors.StorageDealProposal, len(pieces))
+	for i, p := range pieces {
+		sdp := actors.StorageDealProposal{
+			PieceRef:             p.CommP[:],
+			PieceSize:            p.Size,
+			Client:               m.waddr,
+			Provider:             m.maddr,
+			ProposalExpiration:   math.MaxUint64,
+			Duration:             math.MaxUint64 / 2, // /2 because overflows
+			StoragePricePerEpoch: types.NewInt(0),
+			StorageCollateral:    types.NewInt(0),
+			ProposerSignature:    nil, // nil because self dealing
+		}
+
+		deals[i] = sdp
+	}
+
+	params, aerr := actors.SerializeParams(&actors.PublishStorageDealsParams{
+		Deals: deals,
+	})
+	if aerr != nil {
+		return cid.Undef, xerrors.Errorf("serializing PublishStorageDeals params failed: ", aerr)
+	}
+
+	smsg, err := m.api.MpoolPushMessage(ctx, &types.Message{
+		To:       actors.StorageMarketAddress,
+		From:     m.waddr,
+		Value:    types.NewInt(0),
+		GasPrice: types.NewInt(0),
+		GasLimit: types.NewInt(1000000),
+		Method:   actors.SMAMethods.PublishStorageDeals,
+		Params:   params,
+	})
+	if err != nil {
+		return cid.Undef, err
+	}
+
+	return smsg.Cid(), nil
+}
+
+func (m *StorageMinerNodeAdapter) WaitForSelfDeals(ctx context.Context, publicStorageDealsMsgCid cid.Cid) ([]uint64, uint8, error) {
```

Comment:
I think it would be better for this method to get a `statemachine.Context` and send an event async when the deals make it onto the chain. It also allows to send notifications in case the deals ids change in a reorg.